### ISSUE TITLE
How to stop/destroy a subscriber

### ISSUE THREAD

**Issue by picanumber** [sewenew/redis-plus-plus#277]
I have a subscriber that lives in its own thread for the duration of the program. The problem is that following the example for pub/sub given in the documentation, I have no way of letting the subscriber know when to stop consuming (so cannot join the subscriber thread). 

My current way is to set e.g. a 2 sec time limit for responses and use the example method: 

```cpp
// Consume messages in a loop.
while (true) {
    try {
        sub.consume();
    } catch (const Error &err) {
        // Handle exceptions.
        if (exit_flag) break;
    }
}
```
So even if there are no messages to consume, every 2 seconds I'll check whether the flag has been set. Is there a better way to do this ? 

Does the subscriber get a specific message (or generate a specific error) on disconnection? If that's true, I can handle that message/exception and check the flag there + use a specific channel to post my "terminate" request. This way I can guarantee that the subscriber thread will exit even if the server is down.

**Comment by wingunder** [comment]
Hi @picanumber,

There is a related (and closed) issue #240 on this.
Basically you need to set your connection options so that `socket_timeout` is set to some acceptable timeout. Your non-subscribe thread can then set a flag or signal that the subscribe-thread should stop. In the catch-part of the consume() try, you'll then get a `TimeoutError` every time `socket_timeout` is reached.

NOTE: In the example, not all exceptions are caught. Please see the [exception examples in the README](https://github.com/sewenew/redis-plus-plus#examples-1)

Here's a code snippet, showing a possible way to do it in the subscribe-thread.
```c++
#include <atomic>
std::atomic<bool> running(true);

while (running) {
    try {
        sub.consume();
    } catch (const TimeoutError& e) {
       // Do nothing, as we expect a timeout, as we set socket_timeout.
       continue;
    } catch (const Error &e) {
        // Handle errors.
        running = false;
    }
}
```
I hope this helps you.
Regards

**Comment by picanumber** [comment]
Thanks for the reply and the linked info.
